### ISSUE TITLE
Remove hard-coded service IPs

### ISSUE THREAD

**Issue by NicolasT** [scality/metalk8s#780]
In several places service IPs like `10.96.0.10` for cluster DNS are hard-coded. This should be calculated from `BootstrapConfiguration` (or the default service network), and then either stored as a static value in the pillar as well, or stored in the mine, TBD

**Comment by NicolasT** [comment]
This is a release-blocker for https://github.com/scality/metalk8s/milestone/15

**Comment by NicolasT** [comment]
Some clarification: we currently *assume* the `serviceNetwork` is `10.96.0.0/12`:
```
$ git grep '10\.96'
docs/developer/architecture/deployment.rst:  Default: ``10.96.0.0/12``
docs/quickstart/introduction.rst:- A network for :term:`Services <Service>`, defaulting to ``10.96.0.0/12``
salt/_pillar/metalk8s.py:DEFAULT_SERVICE_NETWORK = '10.96.0.0/12'
salt/metalk8s/defaults.yaml:  service_ip: "10.96.0.1"
salt/metalk8s/defaults.yaml:  cluster_dns_ip: 10.96.0.10
salt/metalk8s/kubernetes/kubelet/standalone.sls:          - 10.96.0.10
```

However, the `serviceNetwork` can be configured in `BootstrapConfiguration`, so those addresses would be incorrect if the default `serviceNetwork` is not used.

Hence, we should properly calculate these addresses from the (configured or default, but always available in Pillar) `service_network`, and use them accordingly. We could, in line with what's there now, decide that

- the `kubernetes` *ClusterIP* address is the second address in the `service_network`. I.e., if the `service_network` is `10.96.0.0/12`, then the `kubernetes` *ClusterIP* is `10.96.0.1`
- the `coredns` *ClusterIP* address is the 11th address in the `service_network`. I.e., if the `service_network` is `10.96.0.0/12`, then the `coredns` *CusterIP* is `10.96.0.10`.

Not sure where that calculation should happen. In any case, the 'default' in `salt/metalk8s/default.yaml` has to be gone because we *can't* have a non-parametrized default.

When calculating the address, we must also assert the result *is within the `service_network`*. As an example, if `service_network` is `192.168.0.0/29` we would calculate the `coredns` *ClusterIP* as `192.168.0.10`, but **that's not within the `service_network`**. In that case, we should error out during deployment/SLS rendering/whatever (a `/29` `service_network` is way too small anyway).

**Comment by Ebaneck** [comment]
Todo:

- Obtain the service CIDR from bootstrap config e.g CIDR = 10.96.0.0/12
A user can re-define this value but it needs to be validated in the step below.

- Validate the net mask(/29 and /30 should early return with errors) -> regex get this value or find a python module to calculate the number of host from an ip address slash notation(we need minimum 14 host address).

- write a python method to calculate and set the ClusterIP to be the second address in the service_network
- Write the Kubernetes ClusterIP to the pillars and then remove the hardcoded values replacing them with corresponding pillar values
- write a python method to calculate and set the Coredns ClusterIP to be the 11th address in the service_network
- Write the Coredns clusterIp to the pillars and then replace the hard coded values to corresponding pillar value were necessary.

- Above calculations will be done in `metalk8s.py` pillar file.

**Comment by slaperche-scality** [comment]
> find a python module to calculate the number of host from an ip address slash notation(we need minimum 14 host address).

If we use Python 3, the [ipaddress](https://docs.python.org/3/library/ipaddress.html#prefix-net-mask-and-host-mask) from the stdlib will definitely helps you

**Comment by NicolasT** [comment]
>     * Obtain the service CIDR from bootstrap config e.g CIDR = 10.96.0.0/12
>       A user can re-define this value but it needs to be validated in the step below.

It's already in the pillar, nothing to do here.

>     * Validate the net mask(/29 and /30 should early return with errors) -> regex get this value or find a python module to calculate the number of host from an ip address slash notation(we need minimum 14 host address).

This is not where regexes apply. `/29` and `/30` were just examples (heck, you'd need to check for `/31` and `/32` then as well). The intent was simply: if ever we calculate an address which we enforce as *ClusterIP*, we must at that point also ensure this address is within the given network.


>     * write a python method to calculate and set the ClusterIP to be the second address in the service_network

There's some address calculations stuff in standard Salt modules already.

>     * Write the Kubernetes ClusterIP to the pillars and then remove the hardcoded values replacing them with corresponding pillar values

I'd argue the calculated ones have no place in the pillar. The calculation function is 'pure' and can be abstracted as either a module or a macro. No need to litter pillars with calculated/derived values.

>     * write a python method to calculate and set the Coredns ClusterIP to be the 11th address in the service_network

Same as above.
 
>     * Write the Coredns clusterIp to the pillars and then replace the hard coded values to corresponding pillar value were necessary.

Same as above.

>     * Above calculations will be done in `metalk8s.py` pillar file.

No.

**Comment by NicolasT** [comment]
> > find a python module to calculate the number of host from an ip address slash notation(we need minimum 14 host address).
> 
> If we use Python 3, the [ipaddress](https://docs.python.org/3/library/ipaddress.html#prefix-net-mask-and-host-mask) from the stdlib will definitely helps you

We don't have Python 3 for now. Doing these address calculations is trivial, and there are functions in Salt's `__utils__` (in the `network` module) that likely provide 80% if not 100% of the functionality we need.

**Comment by Ebaneck** [comment]
Todo:
- Obtain the service network from the pillars.
- Compute the `cluster_dns_ip` and `kubernetes_cluster_ip` using `salt.modules.network` module. e.g
```
def compute_service_address():
    cidr = __pillar__.get('networks', {}).get('service')
    if not cidr:
        return
    try:
        cluster_dns_ip = __salt__['network.ip_addrs'](cidr=cidr)[1]
        kubernetes_cluster_ip = __salt__['network.ip_addrs'](cidr=cidr)[10]
    except ValueError as exc:
        return __utils__['pillar_utils.errors_to_dict'](
            'Value error has occurred {}'.format(exc)
        )
    if not cluster_dns_ip and not kubernetes_cluster_ip:
        return __utils__['pillar_utils.errors_to_dict'](
            'Unable to set coredns and kubernetes cluster service IP address' 
        )
    return {
        'cluster_dns_ip': cluster_dns_ip,
        'kubernetes_cluster_ip': kubernetes_cluster_ip,
    }

```
- Use the module within sls files to replace the hard-coded values.
`{% set service_IPs = salt.metalk8s_service_address.compute_service_address() %}`

=> cluster_dns_ip = service_IPs['cluster_dns_ip'] and so on....

**Comment by NicolasT** [comment]
Basically, though the Python code above ain't right:

- Didn't look into it yet, but pretty sure `network.ip_addrs` is *not* to be used anywhere: we're calculating an address from a (given) netrange, this has 0 relation to the system on which the code is running
- `pillar_utils`?
- This code can, quite literally, not fail, unless `networks.service` is not in the pillar for whatever reason

**Comment by Ebaneck** [comment]
> Basically, though the Python code above ain't right:
> 
>     * Didn't look into it yet, but pretty sure `network.ip_addrs` is _not_ to be used anywhere: we're calculating an address from a (given) netrange, this has 0 relation to the system on which the code is running
> 
I would probably argue against this. The documentation had stated "Providing a CIDR via 'cidr="10.0.0.0/8"' will return only the addresses which are within that subnet"
So i am thinking this is clearly independent of the system. No??

>     * `pillar_utils`?
> 
>     * This code can, quite literally, not fail, unless `networks.service` is not in the pillar for whatever reason

I would accept to this, but the extra bits of pieces(try/except) was added to handle failure. As an example, a case where we pass a non-conformant service network like you earlier stated. Something along... 192.168.0.0/29. 

Meaning we would probably not be able to access the `11 element` from the object and that should raise some exception. No??

**Comment by NicolasT** [comment]
> > Basically, though the Python code above ain't right:
> > ```
> > * Didn't look into it yet, but pretty sure `network.ip_addrs` is _not_ to be used anywhere: we're calculating an address from a (given) netrange, this has 0 relation to the system on which the code is running
> > ```
> 
> I would probably argue against this. The documentation had stated "Providing a CIDR via 'cidr="10.0.0.0/8"' will return only the addresses which are within that subnet"
> So i am thinking this is clearly independent of the system. No??

No. The `network.ipaddrs` module gives you **the IP addresses of the minion/host**, and can then optionally give you only the ones in a specific CIDR.

Now, **we don't care about the host/minion's IP addresses** to resolve this issue. To the contrary: if the host/minion *would* have an address that sits within the service network CIDR, then the whole thing would not work at all.

**Comment by Ebaneck** [comment]
> > > Basically, though the Python code above ain't right:
> > > ```
> > > * Didn't look into it yet, but pretty sure `network.ip_addrs` is _not_ to be used anywhere: we're calculating an address from a (given) netrange, this has 0 relation to the system on which the code is running
> > > ```
> > 
> > 
> > I would probably argue against this. The documentation had stated "Providing a CIDR via 'cidr="10.0.0.0/8"' will return only the addresses which are within that subnet"
> > So i am thinking this is clearly independent of the system. No??
> 
> No. The `network.ipaddrs` module gives you **the IP addresses of the minion/host**, and can then optionally give you only the ones in a specific CIDR.
> 
> Now, **we don't care about the host/minion's IP addresses** to resolve this issue. To the contrary: if the host/minion _would_ have an address that sits within the service network CIDR, then the whole thing would not work at all.

yes yes

**Comment by Ebaneck** [comment]
Todo:
- Obtain the service network from the pillars.
- Define a new metalk8s module `metalk8s_network` that will hold network-related utilities
- Add package `python-netaddr` available in all python distributions that gives us possibility to compute addresses on the fly
```
def compute_cluster_address():

    '''
    Determine whether the provided address is within a network range.

    :param service_network (str): CIDR presentation format read from pillars.
     For example, '192.168.1.0/24'.

    :returns Dict: Contains address of cluster dns and kubernetes cluster IP.
        cluster_dns_ip => second object in the list(first usable ip address)
        kubernetes_cluster_ip => 11th object in the list
    '''
    log.info('Obtaining service network from metalk8s pillar')
    cidr = __pillar__.get('networks', {}).get('service')
    if not cidr:
        return
    try:
        ip_list = [ip for ip in IPNetwork(cidr)]
        cluster_dns_ip = str(ip_list[1])
        kubernetes_cluster_ip = str(ip_list[10])
    except ValueError as exc:
        return __utils__['pillar_utils.errors_to_dict'](
            'Network {} is not in CIDR presentation {}'.format(cidr, exc)
        )
    if not cluster_dns_ip and not kubernetes_cluster_ip:
        return __utils__['pillar_utils.errors_to_dict'](
            'Unable to set coredns and kubernetes cluster service IP address'
        )

    return {
        'cluster_dns_ip': cluster_dns_ip,
        'kubernetes_cluster_ip': kubernetes_cluster_ip,
    }
```
- We end up with something like 
```
{'cluster_dns_ip': '192.168.1.1', 'kubernetes_cluster_ip': '192.168.1.10'}

```
- Use the module within sls files to replace the hard-coded values.

**Comment by NicolasT** [comment]
Quick review of the code you pasted above (though we don't *need* that code):

- `if not cidr: return` is a bit odd, just return `None` in an error scenario, while this function claims it returns a `dict` with some fields in it.

- `ip_list = [ip for ip in IPNetwork(cidr)]` constructs a list of roughly 65536 (256**2) strings for a `/16` network, or more than a milion for a `/12` network (which is the size of our default service network), to then pick out the second and 11th element. Seems a bit excessive.

- `if not cluster_dns_ip and not kubernetes_cluster_ip` So, error when *both* are not set. What if one of them is not set?


Anyway, as said, we likely don't need this module at all, and can use a macro (remember we discussed this), and some utilities in Salt's `__utils__['network.*']`:

> there are functions in Salt's __utils__ (in the network module) that likely provide 80% if not 100% of the functionality we need

Looking into the `network` utils, `network_hosts` seems like a very good candidate which does exactly what your custom module does. It *does* construct a whole list, so we may not want that. However, given some other `utils` functions and maybe one more we can get this to work efficiently (and in plain Jinja, no need for any custom modules): create a macro which takes two arguments, a network and an index, and does something like the following:

```
{%
set (netaddress, netsize) = network_cidr.split('/', 1)
set netmask = __utils__['network.cidr_to_ipv4_netmask'](netsize)
set base_address = __utils__['network.get_net_start'](netaddress, netmask)
set base_address_full = __utils__['network.rpad_ipv4_network'](base_address)
set (p1, p2, p3, p4) = base_address_full.split('.')
# IP addresses are just numbers...
set address_number = (p1 * 256 ** 3) + (p2 * 256 ** 2) + (p3 * 256 ** 1) + (p4 * 256 ** 0)

set wanted_address_number = address_number + index
set wanted_address = __utils__['network.hex2ip'](hex(wanted_address_number))

assert __utils__['network.in_subnet'](service_network_cidr, wanted_address)

# At this point, emit `wanted_address` from the macro
%}
```

(not sure how to `assert` in a Jinja macro...)

Then, create two macros, one for each address we need, which invokes the above macro by passing in the network CIDR from the pillar, and the desired index (1 or 11).

Finally, use this macro wherever we need it.

**Comment by Ebaneck** [comment]
> Quick review of the code you pasted above (though we don't _need_ that code):
> 
>     * `if not cidr: return` is a bit odd, just return `None` in an error scenario, while this function claims it returns a `dict` with some fields in it.
ok

> 
>     * `ip_list = [ip for ip in IPNetwork(cidr)]` constructs a list of roughly 65536 (256**2) strings for a `/16` network, or more than a milion for a `/12` network (which is the size of our default service network), to then pick out the second and 11th element. Seems a bit excessive.
> 
>     * `if not cluster_dns_ip and not kubernetes_cluster_ip` So, error when _both_ are not set. What if one of them is not set?

I think if one was not set properly, it should die in the `try` block since we cannot `cast` a null object no??
str(None)  => 'None'
So i am wrong about this.
> 
> 
> Anyway, as said, we likely don't need this module at all, and can use a macro (remember we discussed this), and some utilities in Salt's `__utils__['network.*']`:
> 
Yes, I clearly remember we discussed about this. I maybe had worries down the line since a couple of salt help tips stressed out that `Only use jinja for rendering`

> > there are functions in Salt's **utils** (in the network module) that likely provide 80% if not 100% of the functionality we need
> 
> Looking into the `network` utils, `network_hosts` seems like a very good candidate which does exactly what your custom module does. It _does_ construct a whole list, so we may not want that. However, given some other `utils` functions and maybe one more we can get this to work efficiently (and in plain Jinja, no need for any custom modules): create a macro which takes two arguments, a network and an index, and does something like the following:
> 
> ```
> {%
> set (netaddress, netsize) = network_cidr.split('/', 1)
> set netmask = __utils__['network.cidr_to_ipv4_netmask'](netsize)
> set base_address = __utils__['network.get_net_start'](netaddress, netmask)
> set base_address_full = __utils__['network.rpad_ipv4_network'](base_address)
> set (p1, p2, p3, p4) = base_address_full.split('.')
> # IP addresses are just numbers...
> set address_number = (p1 * 256 ** 3) + (p2 * 256 ** 2) + (p3 * 256 ** 1) + (p4 * 256 ** 0)
> 
> set wanted_address_number = address_number + index
> set wanted_address = __utils__['network.hex2ip'](hex(wanted_address_number))
> 
> assert __utils__['network.in_subnet'](service_network_cidr, wanted_address)
> 
> # At this point, emit `wanted_address` from the macro
> %}
> ```
> 
> (not sure how to `assert` in a Jinja macro...)
> 
> Then, create two macros, one for each address we need, which invokes the above macro by passing in the network CIDR from the pillar, and the desired index (1 or 11).
> 
> Finally, use this macro wherever we need it.

By the way, I can certainly go down this road.

Thanks for the multiple feedback :)

**Comment by charlesprost** [comment]
Let's make sure we backport this to 2.0

**Comment by slaperche-scality** [comment]
Better: it's not merged yet so let's target 2.0 from the start.

**Comment by gdemonet** [comment]
@NicolasT FYI, we can't access `__utils__` or other *dunder* dicts from Jinja macros. We may be able to *run* Salt execution modules within SLS formulas, but defining Jinja macros seems quite limited, as is any code in `.jinja` files such as `map.jinja` (it seems we can reach `salt['grains.filter_by']`, but not all functions), much more limited than what is documented [here](https://docs.saltstack.com/en/latest/topics/jinja/index.html), which seems to only apply to SLS files. Salt, and its magic, are quite confusing... Any *dunder* dict, and `__utils__` in particular, must be accessed from Python code.
